NPU plugin support (#2066)
intel/intel-device-plugins-for-kubernetes: cmd/npu_plugin/README.md

- The NPU KMD source is in the [Linux kernel](https://github.com/torvalds/linux/tree/master/drivers/accel/ivpu).
- Firmware sources are in [linux-firmware](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/intel/vpu) and [intel-npu-driver](https://github.com/intel/linux-npu-driver/tree/main/firmware/bin).

## Pre-built Images

[Pre-built images](https://hub.docker.com/r/intel/intel-npu-plugin)
are available on the Docker hub. These images are automatically built and uploaded
to the hub from the latest main branch of this repository.

Release tagged images of the components are also available on the Docker hub, tagged with their
release version numbers in the format `x.y.z`, corresponding to the branches and releases in this
repository.

See [the development guide](../../DEVEL.md) for details if you want to deploy a customized version of the plugin.

## Installation

There are multiple ways to install Intel NPU plugin to a cluster. The most common methods are described below.

> **Note**: Replace `<RELEASE_VERSION>` with the desired [release tag](https://github.com/intel/intel-device-plugins-for-kubernetes/tags) or `main` to get `devel` images.

> > **Note**: Add ```--dry-run=client -o yaml``` to the ```kubectl``` commands below to visualize the YAML content being applied.

Author: I disagree. This is a note that advises to use these `args` to view the content. Whether it is to be done before or after the actual apply is not relevant.

# Intel NPU device plugin for Kubernetes

Table of Contents

* [Introduction](#introduction)
* [Modes and Configuration Options](#modes-and-configuration-options)
* [Pre-built Images](#pre-built-images)
* [Installation](#installation)
    * [Install with NFD](#install-with-nfd)
    * [Install with Operator](#install-with-operator)
    * [Verify Plugin Registration](#verify-plugin-registration)
* [Testing and Demos](#testing-and-demos)

## Introduction

Intel NPU plugin facilitates Kubernetes workload offloading by providing access to Intel CPU neural processing units supported by the host kernel.

The following CPU families are currently detected by the plugin:
* Core Ultra Series 1
* Core Ultra Series 2

> * Core Ultra 200V Series

Author: I was requested to use the official names. :) But I agree that the codenames are familiar for people, maybe even more familiar.

# Intel NPU device plugin for Kubernetes

Table of Contents

* [Introduction](#introduction)
* [Modes and Configuration Options](#modes-and-configuration-options)
* [Installing driver and firmware for Intel NPUs](#installing-driver-and-firmware-for-intel-npus)
* [Pre-built Images](#pre-built-images)
* [Installation](#installation)
    * [Install with NFD](#install-with-nfd)
    * [Install with Operator](#install-with-operator)
    * [Verify Plugin Registration](#verify-plugin-registration)
* [Testing and Demos](#testing-and-demos)

## Introduction

> Intel NPU plugin facilitates Kubernetes workload offloading by providing access to neural processing unit on Intel CPUs supported by the host kernel.

Reviewer: Make it clearer kernel support refers to NPU, not (just) CPU:
```suggestion
Intel NPU plugin facilitates Kubernetes workload offloading by providing access to  Intel CPU neural processing units supported by the host kernel.
```

* Meteor Lake
* Lunar Lake
* Arrow Lake

Intel NPU plugin may register one resource to the Kubernetes cluster:
| Resource | Description |
|:---- |:-------- |
| npu.intel.com/npu | NPU |

## Modes and Configuration Options

| Flag | Argument | Default | Meaning |
|:---- |:-------- |:------- |:------- |
| -shared-dev-num | int | 1 | Number of containers that can share the same NPU device |
| -allocation-policy | string | none | 3 possible values: balanced, packed, none. For shared-dev-num > 1: _balanced_ mode spreads workloads among NPU devices, _packed_ mode fills one GPU fully before moving to next, and _none_ selects first available device from kubelet. Default is _none_. |

The plugin also accepts a number of other arguments (common to all plugins) related to logging.
Please use the -h option to see the complete list of logging related options.

## Installing driver and firmware for Intel NPUs

In case your host's operating system lacks support for Intel NPUs, see the [NPU driver page](https://github.com/intel/linux-npu-driver) for help.

## Pre-built Images

[Pre-built images](https://hub.docker.com/r/intel/intel-npu-plugin)

> of this component are available on the Docker hub. These images are automatically built and uploaded

Reviewer: Redundant, as this README is about NPU:
```suggestion
are available on the Docker hub. These images are automatically built and uploaded
```

to the hub from the latest main branch of this repository.

Release tagged images of the components are also available on the Docker hub, tagged with their
release version numbers in the format `x.y.z`, corresponding to the branches and releases in this
repository.

See [the development guide](../../DEVEL.md) for details if you want to deploy a customized version of the plugin.

## Installation

There are multiple ways to install Intel NPU plugin to a cluster. The most common methods are described below.

> **Note**: Replace `<RELEASE_VERSION>` with the desired [release tag](https://github.com/intel/intel-device-plugins-for-kubernetes/tags) or `main` to get `devel` images.

> > **Note**: Add ```--dry-run=client -o yaml``` to the ```kubectl``` commands below to visualize the yaml content being applied.

Reviewer: Abbreviation case:
```suggestion
> **Note**: Add ```--dry-run=client -o yaml``` to the ```kubectl``` commands below to visualize the YAML content being applied.
```

### Install with NFD

Deploy GPU plugin with the help of NFD ([Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery)). It detects the presence of Intel NPUs and labels them accordingly. GPU plugin's node selector is used to deploy plugin to nodes which have such a NPU label.

```bash
# Start NFD - if your cluster doesn't have NFD installed yet
$ kubectl apply -k 'https://github.com/intel/intel-device-plugins-for-kubernetes/deployments/nfd?ref=<RELEASE_VERSION>'

> # Create NodeFeatureRules for detecting GPUs on nodes

Reviewer: I do not see any changes to GPU plugin code for detecting NPUs?

Author: Copy-pasta error. I took the GPU plugin as base.

$ kubectl apply -k 'https://github.com/intel/intel-device-plugins-for-kubernetes/deployments/nfd/overlays/node-feature-rules?ref=<RELEASE_VERSION>'

# Create NPU plugin daemonset
$ kubectl apply -k 'https://github.com/intel/intel-device-plugins-for-kubernetes/deployments/npu_plugin/overlays/nfd_labeled_nodes?ref=<RELEASE_VERSION>'
```

### Install with Operator

> NPU plugin can be installed with the Intel Device Plugin Operator. It allows configuring NPU plugin's parameters without kustomizing the deployment files. The general installation is described in the [install documentation](../operator/README.md#installation).

Reviewer: Genitive is redundant:
```suggestion
NPU plugin can be installed with the Intel Device Plugin Operator. It allows configuring NPU plugin parameters without kustomizing the deployment files. The general installation is described in the [install documentation](../operator/README.md#installation).
```
